### DISCUSSION TITLE
"Best Practices for Reusable Bundles" versus Documentation

### DISCUSSION THREAD

**Discussion by chess-server** [symfony/symfony#43117] [Q&A] [answered]
I created a reusable bundle according to the Best Practices section. When loading the bundle, twig does not find the templates under templates, but it would find them under src/Resources/views. 
When I am adding the path 
'%kernel.project_dir%/vendor/acme/name/templates' => 'AcmeName'
in the twig.php config file, it also works. 

I would have expected that this is either noted in the Best Practices or the path is added automatically. 

Or should the Best Practices better be ignored?

**Comment by ahundiak** [comment]
The [views](https://symfony.com/doc/current/bundles.html#bundle-directory-structure) directory is already documented under the bundles section.  I don't see a real need to copy the info to the best practices section.

The name itself dates back to the original release of 2.0. Back then, everything was inside of a bundle.  I think there have been a few proposals to rename it over the years but it's never happened.

By the way, the best practices section itself says that it is perfectly fine to ignore best practices if it suits your needs.

**Comment by chess-server** [comment]
Until now, I created bundles with the structure described in the main documentation. However, my understanding from the changes in Symfony 4 is that the Best Practice reflect what Symfony is heading to in the long run for external bundles. 

As I am in need of several smaller bundles, I created a make:bundle command with the "new" structure to avoid refactoring later. If the plan for external bundles is abandoned, then this is fine with me.

However, with the available description, the Best Practices do not function (templates are not found, assets are not automatically set) or I have missed how to change the configuration easily to get the stuff loaded. Independent from the option to ignore the Best Practice document should explain correctly how to use it...
 
I for myself are moving back to the old structure.

**Comment by jfsenechal** [comment]
I have been using the new structure for a long time and it works.
There must be an error somewhere, don't you have an example bundle?

**Comment by chess-server** [comment]
You can find a sample bundle at https://github.com/chess-server/SampleBundle
You need to load it with composer, set the route in config/routes/. Accessing the controller at localhost:/sample does not work with the twig files im templates, but would do if the template is moved to src/Resources.

**Comment by jfsenechal** [comment]
In your AcmeSampleBundle remove getParent and add getPath 

```php
class AcmeSampleBundle extends Bundle
{
    public function getPath(): string
    {
        return dirname(__DIR__);
    }
}
```

**Comment by chess-server** [comment]
Oops, that was a typo. However, with the getPath function, the app still does not find the twig template under templates. 
Do you know if it is an issue if one mixes old and new bundle structures?

**Comment by jfsenechal** [comment]
What did you have in composer.json at the root of the app

```yaml
"autoload": {
    "psr-4": {
      "App\\": "src/App",
      "Acme\\SampleBundle\\": "src/"
    }
  },
```

**Reply by chess-server** [reply]
I am loading the bundle through composer only, using my private local gitlab installation. Therefore adding the bundle src directory should not be required, shouldn't it? At least this works with the old bundle structure. The line `"Acme\\SampleBundle\\": "src/"` is in the composer.json of the bundle.

**Comment by jfsenechal** [comment]
Oups , that was a typo 

```yaml
"Acme\\SampleBundle\\": "src/Acme/SampleBundle/src",
```
if your folder is in src / ...
I think it should be put in the root composer.json
if it's in the vendor folder then you have your bundle in the "require" area but you need the package to go to packagist.org?

**Comment by chess-server** [comment]
No, composer allows:
`  

"repositories": [
{
            "type": "vcs",
            "url": "git@gitlab.home.xx:SampleBundle"
        },
`
to load without packagist.

**Comment by jfsenechal** [comment]
it's weird I installed it via composer and the template is displayed well

```json
  "require-dev": {
    "acme/sample": "dev-main"
  },
  "repositories": [
    {
      "type": "vcs",
      "url": "git@github.com:chess-server/SampleBundle.git"
    }
  ],
```
```bash
composer up
```
Add DependencyInjection directory and create a extension file to load twig config
```php
//AcmeExtension.php
/**
 * This is the class that loads and manages your bundle configuration.
 *
 * @see http://symfony.com/doc/current/cookbook/bundles/extension.html
 */
final class AcMarcheMercrediExtension extends Extension implements PrependExtensionInterface
{
    private PhpFileLoader $loader;

    public function load(array $configs, ContainerBuilder $containerBuilder): void
    {
        $this->loader->load('services.php');
    }

    /**
     * Allow an extension to prepend the extension configurations.
     */
    public function prepend(ContainerBuilder $containerBuilder): void
    {
        $this->loader = $this->initPhpFilerLoader($containerBuilder);

        foreach (array_keys($containerBuilder->getExtensions()) as $name) {
            switch ($name) {               
                case 'twig':
                    $this->loadConfig('twig');
                    break;               
            }
        }
    }

    protected function loadConfig(string $name): void
    {
        $this->loader->load('packages/'.$name.'.php');
    }

    protected function initPhpFilerLoader(ContainerBuilder $containerBuilder): PhpFileLoader
    {
        return new PhpFileLoader(
            $containerBuilder,
            new FileLocator(__DIR__.'/../../config/'),
            null,
            class_exists(ConfigBuilderGenerator::class) ? new ConfigBuilderGenerator(
                $containerBuilder->getParameter('kernel.cache_dir')
            ) : null
        );
    }
}
```

```php
//twig.php
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {

    $containerConfigurator->extension(
        'twig',
        [
            'form_themes' => ['bootstrap_5_layout.html.twig'],
            'paths' => [
                '%kernel.project_dir%/src/Acme/Sample/templates/' => 'Acme',
            ],
        ]
    );
};
```

**Comment by chess-server** [comment, marked as answer]
I finally got it to work. The twig path is not necessary, but when importing the routes defined through annotations, you need to link 
`@AcmeSampleBundle/**src**/Controller/`  
instead of 
`@AcmeSampleBundle/Controller/`
